added tx20 wind sensor

TheEggi opened the pull request:
## Description:
The tx20 is a simple wind sensor, which can be used to get information about wind speed and direction. 
Sensor can be found here: https://www.amazon.de/Technoline-TX-20/dp/B01HXZ3KLA

The sensor must be connected as follows:

Yellow ground (GND)
Green mass (DTR) -> should also go to GND
Red 3.3V (Vcc)
Black GPIO_2 (TxD) -> this one will be configured.

Description of the protocol:
https://www.john.geek.nz/2011/07/la-crosse-tx20-anemometer-communication-protocol/
http://fabrizio.zellini.org/decoding-la-crosse-tx20-anemometer-with-arduino

Implementations:
https://github.com/arendst/Sonoff-Tasmota/blob/development/sonoff/xsns_35_tx20.ino
https://github.com/bunnyhu/ESP8266_TX20_wind_sensor
http://blog.bubux.de/windsensor-tx20-mit-esp8266/

Detailed description of the sensor can be found here: https://github.com/arendst/Sonoff-Tasmota/pull/3146.


**Related issue (if applicable):** 

**Pull request in [esphomeyaml](https://github.com/OttoWinter/esphomeyaml) with YAML changes (if applicable):** OttoWinter/esphomeyaml# https://github.com/OttoWinter/esphomeyaml/pull/275
**Pull request in [esphomedocs](https://github.com/OttoWinter/esphomedocs) with documentation (if applicable):** OttoWinter/esphomedocs#https://github.com/esphome/ESPHome-Docs/pull/150
## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphomedocs](https://github.com/OttoWinter/esphomedocs).

TheEggi committed:
added tx20 wind sensor

TheEggi committed:
fixed some issues for code generation

TheEggi committed:
fixes for esp32

OttoWinter commented on platformio.ini:
```diff
@@ -1,7 +1,7 @@
 
 [platformio]
-env_default = livingroom
+env_default = tx20sensor
```
Remove this change.

OttoWinter commented on platformio.ini:
```diff
@@ -104,3 +104,19 @@ framework = arduino
+framework = arduino
+lib_deps = ${common.lib_deps}
+build_flags = ${common.build_flags}
+src_filter = ${common.src_filter} +<examples/tx20-sample/tx20-sample.cpp>
```
Please only add one of the esp32/esp8266 versions. Each environment makes the test scripts compile all of esphomelib's code base another time.

TheEggi replied:
Fixed.

OttoWinter commented on platformio.ini:
```diff
@@ -104,3 +104,19 @@ framework = arduino
+framework = arduino
+lib_deps = ${common.lib_deps}
+build_flags = ${common.build_flags}
+src_filter = ${common.src_filter} +<examples/tx20-sample/tx20-sample.cpp>
```
Missing newline at end of file

OttoWinter commented on src/esphomelib/application.h:
```diff
@@ -158,8 +160,7 @@ class Application {
    */
-  LogComponent *init_log(uint32_t baud_rate = 115200,
-                         size_t tx_buffer_size = 512);
+  LogComponent *init_log(uint32_t baud_rate = 115200, size_t tx_buffer_size = 512);
```
Please remove these cleanup changes. It makes it hard to review this PR (and makes merge conflicts that much harder to fix). If you want to fix formatting in this file, please do it in another PR.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+// Based on:
```
Remove this.

thomas-eckerstorfer committed:
removed livingroom/tx20sensor change

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+#include "esphomelib/espmath.h"
+#include "esphomelib/log.h"
+#include "esphomelib/text_sensor/text_sensor.h"
+#include "tx20_component.h"
```
```suggestion
#include "esphomelib/sensor/tx20_component.h"
```

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+const uint8_t TX20_RESET_VALUES = 60;  // seconds
+
+const std::vector<std::string> DIRECTIONS = {"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
+                                             "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"};
```
Don't use `std::vector` here. I believe the compiler will not optimize it away if this integration is not used. Please change to a standard C-style array.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+const char UNIT_KMH[] = "km/h";
+const char ICON_WIND_SPEED[] = "mdi:weather-windy";
+const char ICON_WIND_DIRECTION[] = "mdi:sign-direction";
+const char EMPTY[] = "";
```
Move these ICON and UNIT constants to `sensor.h` and `sensor.cpp`

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+static bool tx20_available = false;
+static float tx20_wind_speed_kmh = 0;
+static uint8_t tx20_wind_direction = 0;
+static bool firstInstance = false;
```
`first_instance`

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+  ESP_LOGV(TAG, "Start TX20Component...");
+  if(firstInstance){
+    ESP_LOGE(TAG, "Multiple TX20 Sensors are not supported!!");
+  }
```
No logging in constructor please.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN(" Pin:", TX20Component::pin_);
```
```suggestion
  LOG_PIN("  Pin:", this->pin_);
```

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+  delayMicroseconds(TX20_BIT_TIME / 2);
+
+  for (int bitcount = 41; bitcount > 0; bitcount--) {
+    uint8_t dpin = (digitalRead(this->pin_->get_pin()));
```
```suggestion
    uint8_t dpin = this->pin_->digital_read();
```

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+  // Must clear this bit in the interrupt register,
+  // it gets set even when interrupts are disabled
+  tx20_wind_direction = tx20_sb;
+  tx20_wind_speed_kmh = float(tx20_sc) * 0.36;
```
```suggestion
  tx20_wind_speed_kmh = float(tx20_sc) * 0.36f;
```

OttoWinter commented on src/esphomelib/sensor/tx20_component.h:
```diff
@@ -0,0 +1,59 @@
+
```
Remove blank line at start (will become important in future release)

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+#endif
+#ifdef ARDUINO_ARCH_ESP8266
+  GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, 1 << this->pin_->get_pin());
+#endif
```
> // Must clear this bit in the interrupt register,
> // it gets set even when interrupts are disabled

Why do we need to clear this flag?

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+      tx20_sf = tx20_sf >> 1 | (dpin << 11);
+    }
+
+    delayMicroseconds(TX20_BIT_TIME);
```
So this interrupt will take 41*1220 microseconds (50 milliseconds) to read the wind sensor? That's not good for a real-time system like an ESP - the WiFi controller will not be able to respond to packets during this time and esphomelib's core loop will not continue. Please try re-writing this into an asynchronous version, see the UART sensors for inspiration.

TheEggi replied:
Any one in particular? Did not quite work out the way as I was intending it to do:
- Intention was to move it to the loop and check the with micros() if the elapsed time is >= as my timeout. Result was not quite working.

OttoWinter commented on src/esphomelib/sensor/tx20_component.h:
```diff
@@ -0,0 +1,59 @@
+using TX20WindDirectionTextSensor = text_sensor::TextSensor;
+
+/// This class implements support for the TX20 WindSpeed+WindDirection sensor.
+class TX20Component : public PollingComponent {
```
Why make this a `PollingComponent` when the sensor is asynchronous (using interrupts) ?

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+  uint8_t chk = (tx20_sb + (tx20_sc & 0xf) + ((tx20_sc >> 4) & 0xf) + ((tx20_sc >> 8) & 0xf));
+  chk &= 0xf;
+
+  if ((chk == tx20_sd) && (tx20_sc < 400)) {  // if checksum seems to be ok and wind speed below 40 m/s
```
Is there a reason wind speeds above 40m/s are discarded?

TheEggi replied:
51.1 is the maximum that is supported by the sensor. I have not seen mine go anywhere close to 40 - so my guess is that it is not working so well in that range.

OttoWinter replied:
Ok so the sensor has a checksum, so that should be enough for checking if everything's alright. Maybe somebody else _has_ conditions where the wind sensor gets above 40 m/s (up to the 51.1 max).

Can we just drop the second check? I think the checksum by itself is just fine.

TheEggi replied:
![image](https://user-images.githubusercontent.com/1206407/52190346-e9f8f080-283e-11e9-9ad9-79337b37467c.png)

Was testing this for half a day - seems to be not good enough. Readded the check again.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+      wind_direction_text_sensor_(new TX20WindDirectionTextSensor(wind_direction_text_name)),
+      pin_(pin) {
+  ESP_LOGV(TAG, "Start TX20Component...");
+  if(firstInstance){
```
You can also set `global_tx20_` here and check if `global_tx20_ != nullptr`

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+
+static bool tx20_available = false;
+static float tx20_wind_speed_kmh = 0;
+static uint8_t tx20_wind_direction = 0;
```
Don't make these global variables. Even though only one of these can exist at a time, please make them class members.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+void TX20Component::update() {
+  if (tx20_available) {
+    ESP_LOGV(TAG, "Updating TX20...");
+    std::string direction = DIRECTIONS.at(tx20_wind_direction);
```
Check if `DIRECTIONS` contains the key first. We do not want to crash the system if we get a faulty reading.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    this->wind_direction_sensor_->publish_state(tx20_wind_direction);
+    this->wind_speed_sensor_->publish_state(tx20_wind_speed_kmh);
+    this->wind_direction_text_sensor_->publish_state(direction);
```
Shouldn't `tx20_available` be reset after this block?

TheEggi replied:
Correct.

thomas-eckerstorfer committed:
added suggestions

thomas-eckerstorfer committed:
changed pollingcomponent to component

thomas-eckerstorfer committed:
fixed compile error

thomas-eckerstorfer committed:
removed delay

thomas-eckerstorfer committed:
removed update method

thomas-eckerstorfer committed:
fix some issues

thomas-eckerstorfer committed:
more fixes

TheEggi committed:
more fixes

TheEggi committed:
added more logs

TheEggi committed:
fixed issue

TheEggi committed:
updated log messages

TheEggi committed:
not working update..

TheEggi committed:
changed tx20 to async

TheEggi committed:
reformat + changes for esp32

TheEggi committed:
removed define

TheEggi committed:
added missing define

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,181 @@
+    this->wind_speed_sensor_->publish_state(tx20_wind_speed_kmh);
+    this->wind_direction_text_sensor_->publish_state(direction);
+  } else {
+    ESP_LOGV(TAG, "Incorrect checksum!");
```
Let's make this a warning

TheEggi replied:
Happens quite a few times -> Added a comment and will keep it as verbose.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,181 @@
+
+ESPHOMELIB_NAMESPACE_END
+
+#endif  // USE_TX20
```
Missing newline at end of file

OttoWinter reviewed (commented):
Great work on making this async 👍 

For the text sensor, please see comment in esphomeyaml - I'd prefer if the text sensor part be split off for now. It's too much of an architecture abuse (and I have the feeling it will bite me in the future - especially with changes to how esphomeyaml will conditionally compile files in the future to improve compilation speed)

thomas-eckerstorfer committed:
removed text sensor and added method to get cardinal directions for easy usage in template

TheEggi committed:
removed text sensor

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,172 @@
+static const uint16_t TX20_MAX_TIME = MAX_BUFFER_SIZE * 1220 + 5000;
+static const uint16_t TX20_BIT_TIME = 1200;
+static const std::string DIRECTIONS[] = {"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
+                                         "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"};
```
Remove `DIRECTIONS` now that there's no text sensor. Also `std::string` has the same issue as `std::vector<>`, it prevents compiler optimizations - it would have needed to be `static const char *DIRECTIONS[]`

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,172 @@
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN(" Pin:", this->pin_);
```
```suggestion
  LOG_PIN("  Pin:", this->pin_);
```

OttoWinter commented on src/esphomelib/sensor/tx20_component.h:
```diff
@@ -0,0 +1,54 @@
+
+ protected:
+  static void pin_change_();
+  void decodeAndPublish_();
```
`lower_snake_case`

OttoWinter commented on src/esphomelib/sensor/tx20_component.h:
```diff
@@ -0,0 +1,54 @@
+#ifdef USE_TX20
+
+#include "esphomelib/sensor/sensor.h"
+#include "esphomelib/text_sensor/text_sensor.h"
```
Remove this

OttoWinter commented on src/esphomelib/defines.h:
```diff
@@ -612,4 +613,13 @@
+  #ifndef USE_SENSOR
+    #define USE_SENSOR
+  #endif
+  #ifndef USE_TEXT_SENSOR
```
Remove this too

OttoWinter reviewed (commented):
Ok, looking quite good now.

TheEggi committed:
updated tx20 sensor

TheEggi committed:
Merge remote-tracking branch 'refs/remotes/esphomelib/dev' into dev

TheEggi committed:
fixed compilation errors after changes

TheEggi committed:
added a comment for the get_wind_cardinal_direction function

thomas-eckerstorfer committed:
readded the check for windspeeds > 40ms

OttoWinter committed:
✏ Remove defines.h entry

defines.h no longer contains that stuff. `ESPHOMEYAML_USE` is now only intended for use with esphomeyaml directly, which sets `USE_SENSOR` automatically.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,173 @@
+  }
+}
+
+uint16_t *TX20Component::buffer_ = new uint16_t[MAX_BUFFER_SIZE];
```
This needs to be allocated in the constructor. Otherwise it might be allocated even when the user isn't using the TX20.

OttoWinter committed:
Update tx20_component.cpp

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,169 @@
+
+void TX20Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TX20...");
+  TX20Component::buffer_ = new uint16_t[MAX_BUFFER_SIZE];
```
buffer_ should be allocated in setup (or constructor). If it is allocated globally every user will lose quite a bit of RAM even if they don't use this integration (plus the allocator on the ESPs is sometimes a bit flakey when it is activated pre user code)

TheEggi replied:
If this is the case it makes sense. I even had the code like that, but was under the impression that the code inside of the USE_TX20 only takes effect when it is in use and therefore moved it down to save a line.

OttoWinter commented on src/esphomelib/sensor/tx20_component.cpp:
```diff
@@ -0,0 +1,169 @@
+    buffer_index_++;
+    return;
+  }
+  uint32_t delay = now - start_time_;
```
Updated this to simplify logic. Because now and start_time_ are unsigned integers, the result will always be positive and the correct number too.

TheEggi replied:
Isn't there still the possibility of an overflow?

OttoWinter committed:
Remove private

OttoWinter reviewed (commented):
Okay, let's get this thing done! :)

I updated the PR a bit now (and left some comments for the changes). I will look later if it compiles correctly and if you agree with the changes I made I would be happy to merge this PR

TheEggi committed:
fixed decoding (no more 400 m/s check needed)

kevank commented:
Is this complete?  Can we use a tx20?

TheEggi commented:
Currently you have to use the branch, but other than that it should be working. Update of the docs will follow on the weekend.

OttoWinter commented:
This moved [here](https://github.com/esphome/esphome/pull/275) (esphome-core is no longer used)
